Features/alignment optimization

MarleneBusch opened the pull request:
# Features/alignment optimization

This branch mainly introduces the `AlignmentOptimizer` that can be used to optimize all 28 heliostat kinematic parameters. There are two optimization methods, one uses the motor positions and the other uses raytracing. Optimizing via the motor positions is significantly faster and also more accurate.

This branch also changed other modules.
- The tutorials were not updated in my last feature branch concerning the device, now the default device is also set there.
- The h5 scenarios were updated to also include the power plant position coordinates.
- The actuator used to have a forward method that had no use, i removed it.
- The kinematic module variable names were not consistent, we used to have `actuator_steps`, `motor_steps`, `actuator_pos`, ... all were referring to the same thing that is named `motor_positions` in PAINT, thats why I chose that name for ARTIST as well.
- A new pytest is introduced checking the util method that converts wgs84 to enu.
- In the heliostat module, the `get_surface_points_and_surface_normals` method was moved from the alignement method to the init. This way it is only called once and not with each new alignment.
- Inplace operations in the kinematic and raytracing were removed to make the modules differentiable.
- A `PAINT_to_surface` converter was added to use PAINT data and create scenarios from it.
- Overall I also found pieces of code that used the math library, I changed that to torch.
 

Fixes #91 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

mcw92 commented on artist/field/actuator_linear.py:
```diff
@@ -100,65 +98,63 @@ def __init__(
+    def steps_to_phi(self, motor_position: torch.Tensor) -> torch.Tensor:
         """
-        Calculate phi (angle) from steps.
+        Calculate phi (angle) from the motor position.
```
Can you maybe explain what phi exactly is, i.e., angle between what?
If this is only a "private" helper function, should we maybe rename it to `_steps_to_phi()`?

MarleneBusch replied:
I dont know much more about phi.
But since phi_0  is the "minimal angle, when in minimum motorposition" , I would assume phi is the angle of the joint?

mcw92 commented on artist/field/actuator_linear.py:
```diff
@@ -100,65 +98,63 @@ def __init__(
+            The calculated angle phi.
         """
-        stroke_length = actuator_pos / self.increment + self.initial_stroke_length
+        stroke_length = motor_position / self.increment + self.initial_stroke_length
```
Can you add explaining comments what these quantities mean? It really looks like you now understand what's going on in ARTIST in detail (we do not anymore 😬 ) so it would be great if you could try to make the code more self-explaining by adding more comments.

MarleneBusch replied:
I added more context to the docstrings.

Also, I renamed radius to pivot_radius and phi_0 to initial_angle, as recommended by Max.

mcw92 commented on artist/field/actuator_linear.py:
```diff
@@ -168,38 +164,14 @@ def angles_to_motor_steps(
-        actuator_steps = (stroke_length - self.initial_stroke_length) * self.increment
-        return actuator_steps
-
-    def forward(
```
According to the [PyTorch documentation](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.forward), the `forward` method of each subclass of `torch.nn.Module` should always be overridden, as it defines the computation performed at every call. However, we should definitely think about what should actually happen in `forward` as this is called upon calling the module instance (including hooks).

MarleneBusch replied:
ActuatorLinear is not actually a direct subclass of `torch.nn.Module`  but a subclass of `Actuator`. 
`Actuator` is indeed a subclass of `torch.nn.Module`.

However this problem then exists in multiple classes in actuator/field/

mcw92 replied:
We have checked again with @Markus-Goetz and there should definitely be `forward()` methods for all child classes of `torch.nn.Module`. However, this means we should carefully think about what we expect to happen if we call an instance of each of the above mentioned classes as this would automatically call the corresponding `forward` method (including hooks).
For now, you could implement those `forward` methods where you already know what should happen and make an issue for those where you don't. Ideally, also tag @Filos1992 because we definitely would need his opinion here. For thos classes, where it is unclear and we do not know yet what we want to happen, you could implement "stubs", like:
```
def forward(self) -> None:
    raise NotImplementedError
```
Or would this break already working functionality?

MarleneBusch replied:
I understand that we need to implement them if torch requires us to do so.
I dont understand how i can figure out what the individual forward methods should do. Therefore i would implement your suggestion
 ```
def forward(self) -> None:
    raise NotImplementedError
```
everywhere.
I cannot recall a single part of the artist code where we just call an instance of such a class.
That is why I think it would break working functionality

mcw92 replied:
I assume you mean would *not* break working functionality in the last sentencen, right? If that's the case, we just leave it like it is. Can you create an issue to document the need to discuss this further?

MarleneBusch replied:
yes, sorry, thats what i meant.
I implemented the forward methods like you suggested everywhere and afterwards i tested the code.
There are no errors in combination with the forward methods. I will open an issue.

mcw92 commented on artist/field/heliostat.py:
```diff
@@ -769,18 +783,66 @@ def set_aligned_surface(
+        torch.Tensor
+            The orientation for the given motor position.
+        """
+        device = torch.device(device)
+        orientation = self.kinematic.motor_positions_to_orientation(
+            motor_positions, device
+        )
+        return orientation
```
```suggestion
        return self.kinematic.motor_positions_to_orientation(
            motor_positions, torch.device(device)
        )
```

MarleneBusch replied:
done

mcw92 commented on artist/field/kinematic_rigid_body.py:
```diff
@@ -141,33 +145,29 @@ def align(
         assert (
             len(self.actuators.actuator_list) == 2
         ), "The rigid body kinematic requires exactly two actuators, please check the configuration!"
```
```suggestion
        if len(self.actuators.actuator_list) != 2:
            raise ValueError(f"The rigid body kinematic requires exactly two actuators but {len(self.actuators.actuator_list)} were specified, please check the configuration!")
```
I'd suggest to replace all asserts with proper exceptions as it is in principle possible to turn off asserts. However, this would also require expanding the tests to also cover each exception.

MarleneBusch replied:
Okay, I changed this part of the code to follow your suggestion.
I also opened an issue to fix all other parts of artist where we are using assert statements. 

Maybe this does not need to be solved right now in this branch?

MarleneBusch replied:
I pushed another version where i exchanged all assterts to ValueErrors except in the tests. I did not yet create all test cases.

MarleneBusch replied:
This breaks some tests. I will fix it

MarleneBusch replied:
The pytests work again now. However the code coverage was lowered due to the if statements, This makes the git checks fail. I dont know what to do about that.

mcw92 commented on artist/field/kinematic_rigid_body.py:
```diff
@@ -237,90 +241,220 @@ def align(
-                    - b * -desired_concentrator_normal[:, 1],
-                    a * -desired_concentrator_normal[:, 1]
-                    + b * -desired_concentrator_normal[:, 2],
+                    a * -desired_concentrator_normal[2:3]
```
Just asking out of curiosity: Why do the slices change, i.e., why do we now only have one dimension? Is this because we now have only one "angle" compared to multiple "angles" before (see variable name)?
Can we also put only 2 instead of 2:3? This should be the same, right?

MarleneBusch replied:
Yes, originally Moritz Code worked slightly different and in his code it was possible to calculate multiple orientation matrices at once.
The way we refactored ARTIST doesn't allow this - i think. This method to calculate the orientation is bound to one instance of the kinematic_rigid_body and a single kinematic_rigid_body instance belongs to a single heliostat. A single heliostat can only align itself according to one orientation matrix at once, this is why we only need to calculate one orientation matrix at once. We might have to rethink this for the multi heliostat support anyways but in this current branch it doesnt make sense to name the variables in the plural and expect multiple angles where there will only be one. That is why i changed it to singular, removed the possibilities for lists and had to adapt the slicing.

Do you want me to change it back or should we accept it like i suggested in this branch and change it to something smart once we agreed on a structure for multiple heliostats?

MarleneBusch replied:
Also we cannot just put 2 instead of 2:3. `desired_concentrator_normal` is 4 dimensional. 

Example:
`desired_concentrator_normal` : `tensor([ 0.0000, -0.7071,  0.7071,  0.0000])`
`desired_concentrator_normal[2:]`  : `tensor([0.7071, 0.0000])`
`desired_concentrator_normal[2:3]` :  `tensor([0.7071])`

mcw92 replied:
Regarding your first comment: Okay, understood. I would suggest to not change it back and leave it the way it is right now.
Regarding your second comment:
What about just:
`desired_concentrator_normal[2] : tensor(0.7071)`
The resulting shape is different from `2:3` (no "[]"), not sure it would work in the code. We can also just leave it the way it is. We were just wondering why that is.

mcw92 commented on artist/field/kinematic_rigid_body.py:
```diff
@@ -237,90 +241,220 @@ def align(
+            device=device,
+        )
+
+        joint_1_rotation = (
```
Can you explain the order of the rotation and translation operations performed here?

MarleneBusch replied:
No this would be something Moritz (or maybe Max) would have to do. I did not create the logic behind the kinematic_rigid_body and dont think it would be efficient if i try to really figure it out.

mcw92 commented on artist/scenario.py:
```diff
@@ -31,6 +30,8 @@ class Scenario:
     Attributes
     ----------
+    power_plant_position : torch.Tensor
+        The position of the power plant in lat, lon, alt.
```
```suggestion
        The position of the power plant as latitude, longitude, altitude.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+
+class AlignmentOptimizer:
+    """
+    Implements an alignment optimizer to find optimal kinematic parameters.
```
```suggestion
    Alignment optimizer to find optimal kinematic parameters.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+    Attributes
+    ----------
+    scenario : Scenario
+        The scenario.
+    optimizer : Optimizer
+        The optimizer.
```
Please be a bit more specific about the scenario being optimized, i.e., explain parameters in more detail.

MarleneBusch replied:
okay

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+    -------
+    optimize()
+        Optimize the kinematic parameters.
+    _optimize_kinematic_parameters_with_motor_positions()
+        Optimize the kinematic parameters using the motor positions.
+    _optimize_kinematic_parameters_with_raytracing()
+        Optimize the kinematic parameters using raytracing (slower).
```
I am not sure whether private methods should appear in a class docstring. I think it's not common practice to do so.

MarleneBusch replied:
okay, I don't know much about that but i removed them

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+        Initialize the alignment optimizer.
+
+        The alignment optimizer optimizes parameters of the rigid body kinematics.
+        These parameters can include the 18 kinematic deviations parameters as well as 5 actuator
```
```suggestion
        These parameters can include the 18 kinematic deviations parameters as well as five actuator
```

MarleneBusch replied:
okay

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+        Parameters
+        ----------
+        tolerance : float
+            The optimzer tolerance.
```
```suggestion
            The optimizer tolerance.
```

MarleneBusch replied:
yes

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+            The optimzer tolerance.
+        max_epoch : int
+            The maximum number of optimization epochs.
+        center_calibration_image: torch.Tensor
+            The center of the calibration flux density.
+        incident_ray_direction: torch.Tensor
```
```suggestion
        center_calibration_image : torch.Tensor
            The center of the calibration flux density.
        incident_ray_direction : torch.Tensor
```

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+            The center of the calibration flux density.
+        incident_ray_direction: torch.Tensor
+            The incident ray direction specified in the calibration.
+        motor_positions: Optional[torch.Tensor]
+            The motor positions specified in the calibration (default is None).
+        device: Union[torch.device, str] = "cuda"
```
```suggestion
        motor_positions : Optional[torch.Tensor]
            The motor positions specified in the calibration (default is None).
        device : Union[torch.device, str] = "cuda"
```

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+        if motor_positions is not None:
+            optimized_parameters, self.scenario = (
+                self._optimize_kinematic_parameters_with_motor_positions(
+                    tolerance,
+                    max_epoch,
+                    center_calibration_image=center_calibration_image,
+                    incident_ray_direction=incident_ray_direction,
+                    motor_positions=motor_positions,
+                    device=device,
```
Can you use keyworded arguments consistently?

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+        else:
+            optimized_parameters, self.scenario = (
+                self._optimize_kinematic_parameters_with_raytracing(
+                    tolerance,
```
Consistent use of keyworded arguments (see above)

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+        Optimize the kinematic parameters using the motor positions.
+
+        This optimizer method optimizes the kinematic parameters by extracting the motor positions
+        and incident ray direction from a specific calibration and using the scence's geometry.
```
```suggestion
        and incident ray direction from a specific calibration and using the scene's geometry.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+            The optimzer tolerance.
+        max_epoch : int
+            The maximum number of optimization epochs.
+        center_calibration_image: torch.Tensor
+            The center of the calibration flux density.
+        incident_ray_direction: torch.Tensor
+            The incident ray direction specified in the calibration.
+        motor_positions: torch.Tensor
```
```suggestion
        center_calibration_image : torch.Tensor
            The center of the calibration flux density.
        incident_ray_direction : torch.Tensor
            The incident ray direction specified in the calibration.
        motor_positions : torch.Tensor
```

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+
+            epoch += 1
+
+        # Align heliostat, reason: scenario will be ready to use for raytracing.
+        # can be removed if we decide to only return the optimized paramters.
```
This sounds like a hidden to do in the code. Which of these options do you prefer? Currently, both the list of optimized parameters and the aligned scenario are returned, right? What would be reason to not return the scenario?

MarleneBusch replied:
I thpught about it and decided it always makes sense to return the scenario

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+            The optimzer tolerance.
+        max_epoch : int
+            The maximum number of optimization epochs.
+        center_calibration_image: torch.Tensor
+            The center of the calibration flux density.
+        incident_ray_direction: torch.Tensor
```
```suggestion
        center_calibration_image : torch.Tensor
            The center of the calibration flux density.
        incident_ray_direction : torch.Tensor
```

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+        epoch = 0
+
+        while loss > tolerance and epoch <= max_epoch:
+            # Align heliostat
```
```suggestion
            # Align heliostat.
```

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+                incident_ray_direction=incident_ray_direction, device=device
+            )
+
+            # Create raytracer
```
```suggestion
            # Create raytracer.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+
+            final_bitmap = raytracer.normalize_bitmap(final_bitmap)
+
+            center = utils.get_center_of_mass(
```
PyCharm gives a warning that `plane_e` and `plane_u` are expceted type `torch.Tensor` but get `float`. Maybe you can double-check this :).

MarleneBusch replied:
Yes they are floats, i corrected this.

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+            final_bitmap = raytracer.normalize_bitmap(final_bitmap)
+
+            center = utils.get_center_of_mass(
+                torch.flip(final_bitmap.T, dims=(0, 1)),
```
Why do we need to flip the bitmap here? This is related to issue #97 and we should definitely find out why that is.

MarleneBusch replied:
Yes I agree, but i cannot do that on my own.

mcw92 replied:
We set the urgency of the corresponding issue to "high" (there are now issue labels for this). Could you please tag all persons who might be helpful in resolving this? Maybe we should set up a joint meeting to discuss this with everyone.

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+            self.optimizer.step()
+            self.scheduler.step(loss)
+
+            if epoch in [max_epoch // 4, 2 * (max_epoch // 4), 3 * (max_epoch // 4)]:
```
What is this condition? You only want to log specific epochs? I find it a bit weird to fix the logging interval if the number of epochs itself can in principle vary.

MarleneBusch replied:
mhhh yes it is weird.
My idea was, that i always want to log 3 epochs, no matter how many max_epochs the user sets.
If the user decides on max_epoch=30, but the AlignmentOptimizer is set to only log every 50th epoch, the user will not get any log statements because 30<50. 

I can change it if you want.

mcw92 replied:
Could we introduce a parameter `num_log` to specify how many log messages a user wants?

```
log_step = max_epoch // num_log  # Calculate logging step.
while loss > tolerance and epoch <= max_epoch:
    ...
    if epoch % log_step == 0:
        log.info(...) 
```

MarleneBusch replied:
done

MarleneBusch commented on artist/field/heliostat.py:
```diff
@@ -51,15 +51,23 @@ class Heliostat(torch.nn.Module):
         The preferred reflection direction for rays reflecting off the heliostat.
+    surface_points
+        The original, unaligned surface points.
+    surface_normals
```
done

MarleneBusch commented on artist/field/kinematic_rigid_body.py:
```diff
@@ -237,90 +241,220 @@ def align(
+
+        orientation = self.incident_ray_direction_to_orientation(
+            incident_ray_direction, device=device
+        ).squeeze()
```
1. orientation = self.incident_ray_direction_to_orientation(incident_ray_direction, device=device) I removed the squeeze here, because it doesnt do anything at all, i guess its left over from previous work.
2. aligned_surface_points = (orientation @ surface_points.unsqueeze(-1)).squeeze(-1) This statement really needs the squeeze and unsqueeze operations because of differing dimensions. orientation.shape = (4, 4), surface_points.shape = (4, 10000, 4) therfore its not possible to just use @ and multiply, an extra dimension has to be introduced and then removed.
3. same for: aligned_surface_normals = (orientation @ surface_normals.unsqueeze(-1)).squeeze(-1)

Instead of squeezing and unsqueezing we can also do:
        aligned_surface_points = surface_points @ orientation.T
        aligned_surface_normals = surface_normals @ orientation.T
The result is the same, I dont know which is better for the runtime

MarleneBusch replied:
done. 

transpose is quicker.

MarleneBusch commented on artist/field/kinematic_rigid_body.py:
```diff
@@ -237,90 +241,220 @@ def align(
+        torch.Tensor
+            The orientation matrix.
+        """
+        assert (
+            len(self.actuators.actuator_list) == 2
+        ), "The rigid body kinematic requires exactly two actuators, please check the configuration!"
```
done

MarleneBusch commented on artist/scenario.py:
```diff
@@ -59,13 +61,16 @@ def __init__(
         Parameters
         ----------
+        power_plant_position : torch.Tensor,
+            The position of the power plant in lat, lon, alt.
```
done

MarleneBusch commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,302 @@
+        Scenario
+            The scenario with aligned heliostat and optimized kinematic parameters.
+        """
+        log.info("Start alignment optimization")
```
done

mcw92 commented on artist/field/heliostat.py:
```diff
@@ -240,9 +252,10 @@ def from_hdf5(
-                prototype_surface is not None
-            ), "If the heliostat does not have individual surface parameters, a surface prototype must be provided!"
+            if prototype_surface is None:
+                raise ValueError(
```
Could you include a test function checking that this `ValueError` appears as expected (so that the codecov warning goes away)?

MarleneBusch replied:
done

mcw92 commented on artist/field/heliostat.py:
```diff
@@ -627,9 +640,10 @@ def from_hdf5(
-                prototype_kinematic is not None
-            ), "If the heliostat does not have an individual kinematic, a kinematic prototype must be provided!"
+            if prototype_kinematic is None:
+                raise ValueError(
```
See above. There is specific functionality in `pytest` to do that, like:
```
with pytest.raises(ValueError): 
    # Call test function
```

MarleneBusch replied:
done

mcw92 commented on artist/field/heliostat.py:
```diff
@@ -735,9 +751,10 @@ def from_hdf5(
-                prototype_actuator is not None
-            ), "If the heliostat does not have individual actuators, an actuator prototype must be provided!"
+            if prototype_actuator is None:
+                raise ValueError(
```
See above, make sure line is covered by tests :).

MarleneBusch replied:
done

mcw92 commented on artist/field/heliostat.py:
```diff
@@ -792,11 +856,11 @@ def set_preferred_reflection_direction(self, rays: torch.Tensor) -> None:
-        assert self.is_aligned, "Heliostat has not yet been aligned."
-
+        if not self.is_aligned:
+            raise ValueError("Heliostat has not yet been aligned.")
```
See above.

MarleneBusch replied:
done

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,296 @@
+        scheduler : Union[_LRScheduler, ReduceLROnPlateau]
+            The learning rate scheduler.
+        """
+        log.info("Create alignment optimizer")
```
```suggestion
        log.info("Create alignment optimizer.")
```

mcw92 commented on artist/util/alignment_optimizer.py:
```diff
@@ -0,0 +1,296 @@
+        incident_ray_direction : torch.Tensor
+            The incident ray direction specified in the calibration.
+        motor_positions : Optional[torch.Tensor]
+            The motor positions specified in the calibration (default is None).
```
```suggestion
            The motor positions specified in the calibration (default is ``None``).
```

mcw92 commented on artist/util/config_dictionary.py:
```diff
@@ -169,3 +174,10 @@
 convert_nurbs_from_normals = "deflectometry"
 """Defines that the NURBS should be learnt using deflectometry data."""
+
+# Constants for WGS84
+# TODO
+# Is this the right place to save this data?
+WGS84_A = 6378137.0  # Major axis in meters
+WGS84_B = 6356752.314245  # Minor axis in meters
+WGS84_E2 = (WGS84_A**2 - WGS84_B**2) / WGS84_A**2  # Eccentricity squared
```
Can you maybe put this as a (global) variable or even just defined constant into the file or function where this is actually used? This only appears once in `convert_wgs84_coordinates_to_local_enu()`.

MarleneBusch replied:
done

mcw92 commented on artist/util/configuration_classes.py:
```diff
@@ -5,6 +5,50 @@
+    Attributes
+    ----------
+    power_plant_position : torch.Tensor
+        The position of the power plant in lat, lon, alt.
```
```suggestion
        The position of the power plant as latitude, longitude, altitude.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/configuration_classes.py:
```diff
@@ -5,6 +5,50 @@
+
+    Methods
+    -------
+    create_power_plant_dict
```
```suggestion
    create_power_plant_dict()
```

MarleneBusch replied:
done

mcw92 commented on artist/util/configuration_classes.py:
```diff
@@ -5,6 +5,50 @@
+        Parameters
+        ----------
+        power_plant_position : torch.Tensor
+            The position of the power plant in lat, lon, alt.
```
```suggestion
            The position of the power plant as latitude, longitude, altitude.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/configuration_classes.py:
```diff
@@ -5,6 +5,50 @@
+
+        Returns
+        -------
+        Dict[str, Any]
```
```suggestion
        dict[str, Any]
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -1,7 +1,12 @@
+import pathlib
+from typing import Any, Union
 
+# from artist.scenario import Scenario
```
Can this be removed?

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    srange : float
+        Slant range (meters).
+    degree : bool
+        Specifies if input is given in degrees or radians.
```
```suggestion
        Whether input is given in degrees or radians.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+def azimuth_elevation_to_enu(
+    azimuth: torch.Tensor,
+    elevation: torch.Tensor,
+    srange: float = 1.0,
```
Can we rename this to `slant_range`?

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    Parameters
+    ----------
+    coordinates_to_transform : torch.Tensor
+        The coordinates in lat, lon, alt that are to be transformed.
```
```suggestion
        The coordinates in latitude, longitude, altitude that are to be transformed.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+        The east offset in meters, north offset in meters, and the altitude difference from the reference point.
+    """
+    device = torch.device(device)
+    # Convert latitude and longitude to radians
```
```suggestion
    # Convert latitude and longitude to radians.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    lat_tower_rad = torch.deg2rad(reference_point[0])
+    lon_tower_rad = torch.deg2rad(reference_point[1])
+
+    # Calculate meridional radius of curvature for the first latitude
```
```suggestion
    # Calculate meridional radius of curvature for the first latitude.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+        1 - config_dictionary.WGS84_E2 * sin_lat1**2
+    )
+
+    # Calculate transverse radius of curvature for the first latitude
```
```suggestion
    # Calculate transverse radius of curvature for the first latitude.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+        (1 - config_dictionary.WGS84_E2 * sin_lat1**2) ** 1.5
+    )
+
+    # Calculate delta latitude and delta longitude in radians
```
```suggestion
    # Calculate delta latitude and delta longitude in radians.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    dlat_rad = lat_tower_rad - lat_rad
+    dlon_rad = lon_tower_rad - lon_rad
+
+    # Calculate north and east offsets in meters
```
```suggestion
    # Calculate north and east offsets in meters.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    device : Union[torch.device, str]
+        The device on which to initialize tensors (default is cuda).
+
+    torch.Tensor
```
```suggestion
    Returns
    -------
    torch.Tensor
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+        The coordinates of the flux density center of mass.
+    """
+    device = torch.device(device)
+    # Calculate center of mass of the bitmap
```
```suggestion
    # Calculate center of mass of the bitmap.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    """
+    device = torch.device(device)
+    # Calculate center of mass of the bitmap
+    x_indices = torch.arange(bitmap.shape[0], device=device)
+    y_indices = torch.arange(bitmap.shape[1], device=device)
+    x_indices, y_indices = torch.meshgrid(x_indices, y_indices, indexing="ij")
+
+    total_mass = bitmap.sum()
+    normalized_bitmap = bitmap / total_mass
+
+    center_of_mass_x = (normalized_bitmap * x_indices).sum()
+    center_of_mass_y = (normalized_bitmap * y_indices).sum()
+    center_of_mass_bitmap = torch.stack([center_of_mass_x, center_of_mass_y], dim=-1)
```
What exactly are x and y? Are they already somehow related to the ENU coordinate system that we use in ARTIST?

MarleneBusch replied:
i changed them to e and u for east and up

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    center_of_mass_y = (normalized_bitmap * y_indices).sum()
+    center_of_mass_bitmap = torch.stack([center_of_mass_x, center_of_mass_y], dim=-1)
+
+    # Construct the coordinates relative to target center
```
```suggestion
    # Construct the coordinates relative to target center.
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    return center_coordinates
+
+
+# The type hint for scenario should be Scenario, but Scenario cannot be imported due to circular imports.
```
```suggestion
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+
+# The type hint for scenario should be Scenario, but Scenario cannot be imported due to circular imports.
+def get_rigid_body_kinematic_parameters_from_scenario(
+    scenario: Any,
```
```suggestion
    scenario: "Scenario",
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -1,7 +1,12 @@
-from typing import Union
+import json
+import pathlib
+from typing import Any, Union
```
```suggestion
from typing import Any, TYPE_CHECKING, Union

if TYPE_CHECKING:
    from artist.scenario import Scenario
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -1,7 +1,12 @@
+import pathlib
+from typing import Any, Union
 
+# from artist.scenario import Scenario
```
```suggestion
```
This can be removed if we do the type hint as stated above.

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    return parameters_list
+
+
+# The type hint for scenario should be Scenario, but Scenario cannot be imported due to circular imports.
```
```suggestion
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+# The type hint for scenario should be Scenario, but Scenario cannot be imported due to circular imports.
+def get_calibration_properties(
+    calibration_properties_path: pathlib.Path,
+    scenario: Any,
```
```suggestion
    scenario: "Scenario",
```

MarleneBusch replied:
done

mcw92 commented on artist/util/utils.py:
```diff
@@ -233,3 +241,385 @@ def translate_enu(
+    ----------
+    calibration_properties_path : pathlib.Path
+        The path to the calibration properties file.
+    scenario : Scenario
```
Please pass the power plant position only instead of the whole Scenario since this is the only attribute used.

MarleneBusch replied:
done

mcw92 commented on tests/data/scenario_generators/generate_scenario_for_test_alignment_optimizer.py:
```diff
@@ -0,0 +1,308 @@
+heliostat_file = pathlib.Path(ARTIST_ROOT) / "tests/data/heliostat_properties.json"
+deflectometry_file = pathlib.Path(ARTIST_ROOT) / "tests/data/deflectometry.h5"
+
+with open(calibration_file, "r") as file:
```
Two general comments for this script:

1.) As far as we can tell, you are loading `PAINT` data here. Since this will probably happen more often, it would be good to create a file `paint_loader.py` or similar and write functions that are designed to load different element, e.g. `load_calibration_properties()` based on a file path. Therefore it would be good if you could adjust all of the loads below to be in external functions.

2.) All the strings used in the keys should be mappings. These mappings are in `PAINT` already - for now I would suggest copying the `paint_mappings` file into ARTIST and importing from there -- in the long term we can consider if making `PAINT` a dependency in `ARTIST` makes sense.

MarleneBusch replied:
done

MarleneBusch commented on artist/util/stral_to_surface_converter.py:
```diff
@@ -437,19 +322,14 @@ def generate_surface_config_from_stral(
+                canting_n[f] = torch.tensor(
+                    facet_header_data[7:10], dtype=torch.float, device=device
                 )
                 number_of_points = facet_header_data[10]
```
done.

While looking at the the binary_extractor in PAINT i noticed something that might be wrong there. 

I think - but dont fully know what else you do in PAINT - that in lines 143-150 confuse e and n for the canting vectors. The canting vectors are not really important and are currently not used in artist, therefore nobody or no test checks if these values make sense.

MarleneBusch replied:
i opened a paint issue for you

mcw92 commented on artist/field/kinematic_rigid_body.py:
```diff
@@ -237,90 +244,226 @@ def align(
+            -1
+        )
+
+        # Option 2:
+        # aligned_surface_points = surface_points @ orientation.T
+        # aligned_surface_normals = surface_normals @ orientation.T
```
What is this?

MarleneBusch replied:
This is the solution to another unresolved conversation from the beginning of this pull request. See above.
Originally you asked why we need to squeeze and unsqueeze. Its because of dimensions. If you dont like squeeze and unsqueeze we can choose this commented option that doesnt use squeeze and unsqueeze.

MarleneBusch replied:
done. 

transpose is quicker.

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+        self.step_size = step_size
+
+    @staticmethod
+    def normalize_evaluation_points_for_nurbs(points: torch.Tensor) -> torch.Tensor:
```
Can we rename `points` to `surface_points`?

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+        )
+        return points_normalized
+
+    def fit_nurbs_surface(
```
This function and the `normalize_evaluation_points_for_nurbs` are identical to the functions with the same names in the `StralToSurfaceConverter`. To avoid code duplicates, we suggest to move them to `util.nurbs`.

Alternatively (even better), we could create a `SurfaceConverter` parent class in `util/surface_converter.py`, which defines this parent class including shared methods. The derived child classes (if necessary at all) could then inherit from this class and overwrite or add methods as needed. However, it may be possible to simply create a surface converter class with common attributes and have different methods defining the conversion to be performed.

MarleneBusch replied:
done, i chose this: create a surface converter class with common attributes and have different methods defining the conversion to be performed.

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+        number_eval_points_n : int
+            The number of evaluation points in the north direction used to generate a discrete surface from NURBS.
+        conversion_method : str
+            The conversion method used to learn the NURBS.
```
Could you list possible options for this parameter in the docstring?

MarleneBusch replied:
done

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+        """
+        Generate a surface configuration from a ``PAINT`` dataset.
+
+        Parameters
```
In the long term, we could consider moving the NURBS "base" parameters as attributes to the SurfaceConverter and only pass conversion-dependent parameters here.

MarleneBusch replied:
done

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+        log.info(
+            "Beginning generation of the surface configuration based on PAINT data."
+        )
+        with open(self.heliostat_file_path, "r") as file:
```
Here you are again loading PAINT data. Please move this into a function as mentioned above in the `paint_loader.py` and use mappings instead of strings.

MarleneBusch replied:
done

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+                )
+
+        # Reading ``PAINT`` deflectometry hdf5 file.
+        # TODO
+        # How to deal with different amount of points on different facets?
```
This To Do should be solved before mergin. As mentioned in the STRAL converter, for now, just take the first N points, where N is the number of points on the facet with the smallest number of points.

MarleneBusch replied:
done

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+
+        log.info("Loading PAINT data complete")
+
+        # Select only selected number of points to reduce compute.
```
Once the surface points, surface normals, and facets etc. have been loaded from PAINT or STRAL, this post-processing is identical and should therefore also be moved into a separate function that both converters call.

MarleneBusch replied:
done

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+                    canting_n=canting_n[i],
+                )
+            )
+        log.info("Surface configuration based on STRAL data complete!")
```
Please also adapt the log messages accordingly.

MarleneBusch replied:
done

mcw92 commented on artist/util/paint_to_surface_converter.py:
```diff
@@ -0,0 +1,403 @@
+        with h5py.File(self.deflectometry_file_path, "r") as file:
+            surface_points_with_facets = torch.empty(0, device=device)
+            surface_normals_with_facets = torch.empty(0, device=device)
+            for f in range(number_of_facets):
```
In this loop, you could probably extract the facet with the smallest number of points.

mcw92 replied:
You could also directly extract the surface points / surface normals with list comprehension into "number of facets" torch tensors in the list. Then, you could find the one with the least number of points and adapt the others accordingly.

MarleneBusch replied:
done

mcw92 commented on tests/data/scenario_generators/generate_scenario_for_test_alignment_optimizer.py:
```diff
@@ -0,0 +1,308 @@
+        power_plant_position,
+        device=device,
+    )
+    plane_e = (
```
This looks like a utils function for creating a plane based on corner points.

MarleneBusch replied:
i put it into utils

mcw92 commented on tests/data/scenario_generators/generate_scenario_for_test_alignment_optimizer.py:
```diff
@@ -0,0 +1,308 @@
+    second_joint_tilt_n=torch.tensor(0.0, device=device),
+    second_joint_tilt_u=torch.tensor(0.0, device=device),
+    concentrator_translation_e=torch.tensor(0.0, device=device),
+    concentrator_translation_n=torch.tensor(-0.17755, device=device),
+    concentrator_translation_u=torch.tensor(-0.4045, device=device),
```
These parameters should also now be in PAINT.

MarleneBusch replied:
Not yet in their final form...

MarleneBusch replied:
done

mcw92 commented on tests/field/test_kinematic_with_measured_parameters.py:
```diff
@@ -0,0 +1,196 @@
+        0
+    ].kinematic.deviation_parameters.concentrator_tilt_n = torch.tensor(
+        0.0, device=device
+    )
+    scenario.heliostats.heliostat_list[
+        0
+    ].kinematic.deviation_parameters.concentrator_tilt_u = torch.tensor(
+        0.0, device=device
+    )
+
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        0
+    ].increment = torch.tensor(154166.666, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        0
+    ].initial_stroke_length = torch.tensor(0.075, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        0
+    ].offset = torch.tensor(0.34061, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        0
+    ].pivot_radius = torch.tensor(0.3204, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        0
+    ].initial_angle = torch.tensor(-1.570796, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        1
+    ].increment = torch.tensor(154166.666, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        1
+    ].initial_stroke_length = torch.tensor(0.075, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        1
+    ].offset = torch.tensor(0.3479, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        1
+    ].pivot_radius = torch.tensor(0.309, device=device)
+    scenario.heliostats.heliostat_list[0].kinematic.actuators.actuator_list[
+        1
+    ].initial_angle = torch.tensor(0.959931, device=device)
```
Shouldn't all these parameters already be set in the scenario definition?

MarleneBusch replied:
Max wanted it exactly like this

MarleneBusch replied:
done

mcw92 commented on tests/field/test_kinematic_with_measured_parameters.py:
```diff
@@ -0,0 +1,196 @@
+    device: torch.device,
+) -> None:
+    """
+    Test the kinematic with measured parameters.
```
What is the purpose of this test? It appears to call functions that are already covered in other tests with the only difference being that very specific parameters are set. Shouldn't we have unit tests that already guarantee that these parameters work as desired?

MarleneBusch replied:
We do.

MarleneBusch replied:
done

mcw92 commented on tests/field/test_kinematic_with_measured_parameters.py:
```diff
@@ -0,0 +1,196 @@
+
+    final_bitmap = raytracer.normalize_bitmap(final_bitmap)
+
+    assert final_bitmap.sum() != 0
```
Can you explain this final assert? It looks like we do not check whether the result is really correct but only if "something" happens.

MarleneBusch replied:
yes

MarleneBusch replied:
done

MarleneBusch commented on tests/data/scenario_generators/generate_scenario_for_test_alignment_optimizer.py:
```diff
@@ -0,0 +1,308 @@
+)
+
+# Include actuator parameters for both actuators.
+actuator1_parameters = ActuatorParameters(
+    increment=torch.tensor(154166.666, device=device),
+    initial_stroke_length=torch.tensor(0.075, device=device),
+    offset=torch.tensor(0.34061, device=device),
+    pivot_radius=torch.tensor(0.3204, device=device),
+    initial_angle=torch.tensor(-1.570796, device=device),
+)
```
not yet...

MarleneBusch replied:
done

mcw92 commented on tests/util/test_alignment_optimizer_methods.py:
```diff
@@ -0,0 +1,164 @@
+        threshold_mode="abs",
+    )
+
+    # Load the calibration data
```
```suggestion
    # Load the calibration data.
```

MarleneBusch replied:
done

mcw92 commented on tests/util/test_alignment_optimizer_methods.py:
```diff
@@ -0,0 +1,164 @@
+        ),
+    ],
+)
+def test_alignment_optimizer_methods(
```
This looks like a good integration test but would it be also possible to test the individual parts of the optimizer with unit tests?

mcw92 replied:
I feel like this test only helps to identify problems occurring in the future and does not clearly show that the currently implemented logic of the optimizer works properly.

MarleneBusch replied:
there is no other way to test the alignemnt optimizer methods.

MarleneBusch commented on tests/data/scenario_generators/generate_scenario_for_test_integration_stral.py:
```diff
@@ -22,17 +22,25 @@
 
 # The following parameter is the name of the scenario.
-file_path = "test_scenario"
+file_path = file_path = pathlib.Path(ARTIST_ROOT) / "tests/data/test_scenario_stral"
```
done

mcw92 commented on tests/util/test_alignment_optimizer_methods.py:
```diff
@@ -0,0 +1,164 @@
+        )
+    )
+
+    # Create alignment optimizer
```
```suggestion
    # Create alignment optimizer.
```

MarleneBusch replied:
done

mcw92 commented on tests/util/test_alignment_optimizer_methods.py:
```diff
@@ -0,0 +1,164 @@
+        scheduler=scheduler,
+    )
+
+    if optimizer_method == "use_raytracing":
```
"use_raytracing" should also be a mapping. In general, concrete strings should be avoided.

MarleneBusch replied:
done

mcw92 commented on tests/util/test_wgs84_to_enu.py:
```diff
@@ -0,0 +1,75 @@
+@pytest.mark.parametrize(
+    "wgs84_coordinates, reference_point, expected_enu_coordinates",
+    [
+        # Coordinates of Jülich power plant and multifocustower
```
```suggestion
        # Coordinates of Jülich power plant and multi-focus tower
```

MarleneBusch replied:
done

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+        / "tutorials/data/test_scenario_alignment_optimization.h5"
+    )
+
+# Set up logger
```
```suggestion
# Set up logger.
```

MarleneBusch replied:
done

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+# Set up logger
+set_logger_config()
+
+# Set the device
```
```suggestion
# Set the device.
```

MarleneBusch replied:
done

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+        scenario_file=scenario_file, device=device
+    )
+
+# Choose calibration data
```
```suggestion
# Choose calibration data.
```

MarleneBusch replied:
done

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+    pathlib.Path(ARTIST_ROOT) / "tutorials/data/test_calibration_properties.json"
+)
+
+# Load the calibration data
```
```suggestion
# Load the calibration data.
```

MarleneBusch replied:
done

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+    )
+)
+
+# Get optimizable parameters. (This will choose all 28 kinematic parameters)
```
```suggestion
# Get optimizable parameters. This will select all 28 kinematic parameters.
```

MarleneBusch replied:
done

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+    scenario=example_scenario
+)
+
+# Set up optimizer and scheduler parameters
```
```suggestion
# Set up optimizer and scheduler parameters.
```

MarleneBusch replied:
done

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+learning_rate_patience = 20
+learning_rate_threshold = 0.1
+
+# If you want alignment optimization with raytracing set motor_positions to None and adapt optimizer and scheduler parameters.
+# Uncomment for alignment optimization with raytracing
+# motor_positions = None
+# tolerance = 1e-7
+# max_epoch = 27
+# initial_learning_rate = 0.0002
+# learning_rate_factor = 0.1
+# learning_rate_patience = 18
+# learning_rate_threshold = 0.1
```
I'd prefer no commented-out lines of code in the tutorial. Can we introduce a boolean flag for this?

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+
+optimizer = torch.optim.Adam(parameters, lr=initial_learning_rate)
+
+# Set up learning rate scheduler
```
```suggestion
# Set up learning rate scheduler.
```

MarleneBusch replied:
done

mcw92 commented on tutorials/alignment_optimization_tutorial.py:
```diff
@@ -0,0 +1,98 @@
+    threshold_mode="abs",
+)
+
+# Create alignment optimizer
```
```suggestion
# Create alignment optimizer.
```

MarleneBusch replied:
done

mcw92 reviewed (changes requested):
Thank you for your effort in putting together this code; it was obviously a lot of work.

We’ve provided comments at various points which should be addressed before merging. In addition, the main areas for improvement are as follows:
 1. Ensure that all raise blocks are covered by tests.
 2. Avoid code duplication where possible — there are several instances that would benefit from refactoring.
 3. Please ensure there are no hardcoded PAINT-specific strings in the dictionaries. This functionality should be handled through mapping.
 4. The surface converter could use refactoring. We’ve added multiple comments to guide you in this area.
 5. Finally, aim to include more unit tests. Currently, the tests seem predominantly integration-focused, and unit tests would improve coverage and reliability.

MarleneBusch commented on tests/util/test_wgs84_to_enu.py:
```diff
@@ -0,0 +1,75 @@
+    device: torch.device,
+) -> None:
+    """
+    Test the WGS84 to enu conversion.
```
done

MarleneBusch commented on tests/util/test_wgs84_to_enu.py:
```diff
@@ -0,0 +1,75 @@
+from artist.util import utils
+
+
+@pytest.fixture(params=["cpu", "cuda"] if torch.cuda.is_available() else ["cpu"])
```
done

MarleneBusch commented:
I added a lot of tests to cover all raise blocks.
I merged the stral and paint to surface converters into a single surface converter.
I removed all strings to read paint data and used mapping instead.

kalebphipps reviewed (changes requested):
The number of changed files seems to keep exponentially growing! This is definitely a huge contribution!

Thanks for all the changes - especially all the tests to cover errors and the additional unit tests make a huge difference. I also like the way the `SurfaceConverter` is refactored.

I have included a few comments - most are minor points regarding DocStrings or things that we should maybe note as Issues to deal with later. I think the two important things are:
- A small suggestion to further refactor the `SurfaceConverter`.
- Some notes on files being included from test coverage. I know they were previously excluded, but looking through the code I think there might still be some things we can test there.

I also noticed that the updated keys for the ``PAINT`` data have not yet been included - we should probably update these (i.e. `sun_azimuth` instead of `Sun_azimuth`) and fix the test data before merging right? 

Locally (I will push now) I have formatted the test `.json` files to be "pretty" and also included a numbering for the alignment tutorial in the file name.

Thanks again for all the work - and sorry for the further requested changes...there are so many changes that I keep noticing new things when I review it again...

MarleneBusch commented on artist/util/config_dictionary.py:
```diff
@@ -169,3 +165,101 @@
+"""Key to access the focal spot of a calibration image in a ``PAINT`` calibration.json file."""
+paint_utis = "UTIS"
+"""Key to access the UTIS focal spot in a ``PAINT`` calibration.json file."""
+paint_sun_azimuth = "Sun_azimuth"
```
done

MarleneBusch commented on artist/raytracing/rays.py:
```diff
@@ -33,13 +33,14 @@ def __init__(
+            or (ray_directions.size(dim=1) != ray_magnitudes.size(dim=1))
+            or (ray_directions.size(dim=2) != ray_magnitudes.size(dim=2))
+        ):
+            raise ValueError("Ray directions and magnitudes have differing sizes!")
```
done

MarleneBusch commented on artist/util/surface_converter.py:
```diff
@@ -0,0 +1,625 @@
+        torch.Tensor, torch.Tensor, torch.Tensor, list[torch.Tensor], list[torch.Tensor]
+    ]:
+        """
+        Generate a surface configuration from a ``STRAL`` file.
```
done

MarleneBusch commented on artist/util/surface_converter.py:
```diff
@@ -0,0 +1,625 @@
+        torch.Tensor, torch.Tensor, torch.Tensor, list[torch.Tensor], list[torch.Tensor]
+    ]:
+        """
+        Generate a surface configuration from a ``PAINT`` dataset.
```
done

MarleneBusch commented on artist/util/utils.py:
```diff
@@ -233,3 +239,473 @@ def translate_enu(
+    return parameters_list
+
+
+def normalize_points(points: torch.Tensor) -> torch.Tensor:
```
it is not being tested because its only used in the paint loader. This will be resolved once i write a test for the paint loader

MarleneBusch replied:
done

MarleneBusch commented on artist/util/utils.py:
```diff
@@ -233,3 +239,473 @@ def translate_enu(
+    return points_normalized
+
+
+def corner_points_to_plane(
```
it is not being tested because its only used in the paint loader. This will be resolved once i write a test for the paint loader

MarleneBusch replied:
done

MarleneBusch commented on artist/util/surface_converter.py:
```diff
@@ -0,0 +1,625 @@
+
+        return nurbs_surface
+
+    def generate_surface_config(
```
done

MarleneBusch commented on tests/data/calibration_properties.json:
As of now i dont think I can do that in VSCode this quick, so it would be nice if you keep doing that in the future.

MarleneBusch commented on tests/field/test_set_preferred_reflection_direction.py:
```diff
@@ -0,0 +1,35 @@
+    mocker: MockerFixture,
+) -> None:
+    """
+    Test the heliostat load from hdf5 method.
+
```
done

MarleneBusch commented on tests/scene/test_load_light_source_array.py:
```diff
@@ -0,0 +1,57 @@
+    device: torch.device,
+) -> None:
+    """
+    Test the heliostat load from hdf5 method.
```
done

MarleneBusch commented on tests/util/test_all_forward.py:
i created another issue

MarleneBusch commented on pyproject.toml:
```diff
@@ -133,6 +134,7 @@ testpaths = [
 omit = [
     "artist/util/scenario_generator.py",
-    "artist/util/stral_to_surface_converter.py",
+    "artist/util/surface_converter.py",
```
done

MarleneBusch commented on pyproject.toml:
```diff
@@ -133,6 +134,7 @@ testpaths = [
     "artist/util/scenario_generator.py",
-    "artist/util/stral_to_surface_converter.py",
+    "artist/util/surface_converter.py",
+    "artist/util/paint_loader.py",
```
done

kalebphipps reviewed (approved):
I have made a few changes to the docstrings and test coverage locally which I will now push.

It looks good to me though - I am happy to allow the merge!
